python/src/scipp/compat/xarray_compat.py

```diff
@@ -0,0 +1,97 @@
+import typing
+
+import xarray
+import scipp
+
+
+def from_xarray(xarray_obj: typing.Union[xarray.DataArray, xarray.Dataset]):
+    """
+    Convenience method to convert an xarray object into the corresponding
+    scipp object.
+
+    If you know in advance what type of object you need to convert, you can
+    also call from_xarray_dataarray or from_xarray_dataset directly.
+
+    :param xarray_obj: the xarray object to convert; must be either an xarray
+        DataArray or Dataset object.
+    :return: the converted scipp object.
+    """
+    if isinstance(xarray_obj, xarray.DataArray):
+        return from_xarray_dataarray(xarray_obj)
+    elif isinstance(xarray_obj, xarray.Dataset):
+        return from_xarray_dataset(xarray_obj)
+    else:
+        raise ValueError(
+            f"from_xarray: cannot convert type '{type(xarray_obj)}'")
+
+
+def from_xarray_dataarray(
+        xarray_dataarray: xarray.DataArray) -> scipp.DataArray:
+    """
+    Converts an xarray.DataArray object to a scipp.DataArray object.
+
+    :param xarray_dataset: an xarray.DataArray object to be converted.
+    :return: the converted scipp DataArray object.
+    """
+
+    sc_coords = {}
+    sc_attribs = {}
+
+    for coord in xarray_dataarray.coords:
+        sc_coords[coord] = scipp.Variable(
+            dims=[coord],
+            values=xarray_dataarray.coords[coord].values,
+        )
+
+    for attrib in xarray_dataarray.attrs:
+        sc_attribs[attrib] = scipp.scalar(xarray_dataarray.attrs[attrib])
+
+    return scipp.DataArray(data=scipp.Variable(values=xarray_dataarray.values,
+                                               dims=xarray_dataarray.dims),
+                           coords=sc_coords,
+                           attrs=sc_attribs,
+                           name=xarray_dataarray.name or "")
+
+
+def from_xarray_dataset(xarray_dataset: xarray.Dataset,
+                        *,
+                        attrib_prefix="attrib_") -> scipp.Dataset:
+    """
+    Converts an xarray.Dataset object to a scipp.Dataset object.
+
+    :param xarray_dataset: an xarray.Dataset object to be converted.
+    :param attrib_prefix: customisable prefix for xarray attributes.
+        xarray attributes are stored as scalar values in scipp with
+        a name corresponding to a prefix + the attribute name in
+        xarray.
+
+    :return: the converted scipp dataset object.
+    """
+    sc_data = {}
+    sc_coords = {}
+
+    for data_key in xarray_dataset.data_vars:
+        sc_data[data_key] = from_xarray_dataarray(
+            xarray_dataset.data_vars[data_key])
+
+    for coord in xarray_dataset.coords:
+        sc_coords[coord] = scipp.Variable(
+            dims=[coord],
+            values=xarray_dataset.coords[coord].values,
+        )
```

Comment:
This (make variable from xarray object) is repeated a couple of times, also in `from_xarray_dataarray` above. Can we use common code? xarray also has a `Variable` type, but I am not sure that is what you get from `xarray_dataset.coords[coord]` (that might be a `DataArray`), so I don't know if you can automatically do this using `from_xarray`, or need a helper such as `_from_xarray_variable`:

```python
coords = { key:_from_xarray_variable(var) for key, var in ds.coords.items() }
```
(ideally we'd want to simply use `coords = { key:from_xarray(var) for key, var in ds.coords.items() }`, but you will have to see if you can figure out automatically that it should create a `Variable` and not a `DataArray`).